### PR TITLE
Event dispatcher

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@PVince81 @MorrisJobke @Raydiation @PVince81 @Raydiation

### CODE REVIEW

**Comment by PVince81** [comment]
Looks nice.

We have a few other places in core where we'd need that.

By the way: in what way would this affect the assets pipeline ? Would there be two JS bundles one with the extra scripts and one without ?

**Comment by PVince81** [comment]
We could actually dispatch a similar event after "app.php" of any app has run. Not sure about the performance though if we do it for each app in every request...

**Comment by DeepDiver1975** [comment]
> By the way: in what way would this affect the assets pipeline ? Would there be two JS bundles one with the extra scripts and one without ?

Not at all from my understanding - all scripts/styles will be summed up during rendering of the template. And the version hash of the assembled files will change based on the files included.

**Comment by Xenopathic** [comment]
Death to the hooks! I like this :smile:

File: apps/files/index.php [hunk 1]
```diff
@@ -143,6 +143,17 @@ function renderScript($appName, $scriptName) {
 OCP\Util::addscript('files', 'files');
 OCP\Util::addscript('files', 'navigation');
 OCP\Util::addscript('files', 'keyboardshortcuts');
+
+\OC::$server->getEventDispatcher()->dispatch('OCA\Files::loadAdditionalScripts');
+
+\OC::$server->getEventDispatcher()->addListener('OCA\Files::loadAdditionalScripts', function() {
```
Comment (Xenopathic) [inline apps/files/index.php]: I'm guessing this is example code and isn't actually supposed to be run? Otherwise it seems odd the dispatch call is above

**Comment by DeepDiver1975** [comment]
> We could actually dispatch a similar event after "app.php" of any app has run. Not sure about the performance though if we do it for each app in every request...

We have to do this slightly different
1. load all apps
2. emit the app-loaded hook for all of them

Otherwise an app being loaded later cannot react on a loaded event of a previous app

**Comment by PVince81** [comment]
Anyway, I'm fine with the current approach for now.
Then we can adapt all files_* apps to load their scripts in that handler instead of app.php.

File: apps/files/index.php (hunk 1 above)
Comment (MorrisJobke) [inline apps/files/index.php]: I don't get it. In this case the dispatch happens before the registration of the listeners. So the listener isn't called. Or am I totally wrong and missunderstood the EventDispatcher?

Comment (PVince81) [inline apps/files/index.php]: @MorrisJobke this block is supposed to be in the activity app's appinfo/app.php.
@DeepDiver1975 has added it only to show an example. That must be removed before merging.

Comment (MorrisJobke) [inline apps/files/index.php]: I guess this will completely mess up once the apps are not loaded in the correct order.

Comment (DeepDiver1975) [inline apps/files/index.php]: sure - this will go into the activity app

**Comment by Xenopathic** [comment]
:+1:

**Comment by Raydiation** [comment]
:+1:

**Comment by PVince81** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
Let me squash the last two commits later before merge ...

**Comment by PVince81** [comment]
@DeepDiver1975 time to squash

**Comment by PVince81** [comment]
Please fix the commit text

**Comment by PVince81** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@owncloud-bot :alien:

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Pull request merged**